### PR TITLE
support providing `project-name` via hooks, env vars and template value file

### LINKED ISSUE #522: support providing `project-name` via hooks, env vars and template value file
Hi, 
I wanted to create a template that allows me to create a project within URL. The question about `project-name` in most cases is in GitHub link and this is in my opinion, the best approach in creating a new project from template. The `project-name` variable must be required, but checking about its absence should be postponed until all pre-hooks are called.

This is a pseudo dialogue I want to achieve:

**Past link to repository of your project:** <github.com/organizaton/my-awesome-rust-project>
**Is the name of your project correct `my-awesome-rust-project`?** [Y (Yes)/e (Edit)] 
**Your project is created!!**

To make this backward compatible the template should set options that allow it to set `project-name` on it's own. For example in `cargo-generate.toml` `template.project-name.ask = false` (default `true`)

What do you think about this idea?

### LINKED ISSUE #637: feat: explain when pre/post scripts are run
## PRE
* Currently pre hooks are run after all placeholders mentioned in `cargo-generate.toml` has been resolved.
* The hooks are free to add aditional variables, but its too late to influence the conditional system.

This is a sideeffect of conditionals influencing the hooks - so placeholders need to be evaluated before the hooks are known.

## POST
* Post scripts are run *after* template expansion, but *before* final output is moved to the final destination.

Why not later? Security, and the fact that a failing scrips still causes no errors in the users destination.

### PR DESCRIPTION
**Pull request by taurr**
Fixes #522
Fixes #637

### CODE REVIEW

**Comment by taurr** [comment]
@sassman I could use a little help here - don't know why it fails on windos and mac - and I don't have access to either atm. :-(

**Comment by taurr** [comment]
Had a hunch, tried it - and it worked...

I do not *understand* why it was a problem though, and that bothers my slightly autistic brain though!

**Comment by taurr** [comment]
@sassman have you left us dear friend? Still busy with real life?

**Comment by sassman** [comment]
> @sassman have you left us dear friend? Still busy with real life?

@taurr just taking a break and traveling at the moment. BRB in mid September.

**Comment by taurr** [comment]
@sassman what's the status on this?

on another note: gratz with the new surname :-)

**Comment by sassman** [comment]
> on another note: gratz with the new surname :-)

Thanks a lot, as you can guess a lot of things have been going on.

> @sassman what's the status on this?

So I hope that things will settle a bit in the next week and I hope to have it reviewed soon.
Also, a release is more than overdue.

**Comment by taurr** [comment]
@sassman all ready - please take a look

File: src/lib.rs [hunk 1]
```diff
@@ -110,8 +96,9 @@ fn internal_generate(mut args: GenerateArgs) -> Result<PathBuf> {
     let mut config = Config::from_path(
         &locate_template_file(CONFIG_FILE_NAME, &template_base_dir, &template_dir).ok(),
     )?;
-    config.template.get_or_insert(Default::default());
 
+    // the `--init` parameter may also be set by the template itself
+    // TODO: should we show a warning to the user if the template doesn't agree with the parameter? The user might not expect the template files to end up in CWD!
```
Comment (sassman) [inline src/lib.rs:101]: @taurr shall we create a followup tracking issue for that `TODO` maybe?

File: src/template.rs [hunk 2]
```diff
@@ -30,43 +32,56 @@ fn engine() -> liquid::Parser {
...
+    project_dir: &ProjectDir,
+    project_name: &ProjectName,
+    crate_name: &CrateName,
+) -> Result<()> {
+    liquid_object.insert(
+        "project-name".into(),
+        Value::Scalar(project_name.as_ref().to_owned().into()),
+    );
+    liquid_object.insert(
+        "crate_name".into(),
```
Comment (sassman) [inline src/template.rs:73]: when I see `project-name`, `crate_name` and `within_cargo_project` in this change set here, I just realized how inconsistent we got with the naming convention of dash-case or snake_case.

But it's not a big deal, just some thing we might be able to fix in the future with either/or

**Review by sassman — APPROVED** [review]
Looks very good, thanks for your hard work and patience.

File: src/lib.rs (hunk 1 above)
Comment (taurr) [reply src/lib.rs:101]: #763

**Pull request merged**